aten/src/THC/THCAtomics.cuh

```diff
@@ -40,7 +40,8 @@ struct AtomicAddIntegerImpl<T, 2> {
 
     do {
       assumed = old;
-      sum = val + (size_t)address & 2 ? T(old >> 16) : T(old & 0xffff);
+      sum = (size_t)address & 2 ? old >> 16 : old & 0xffff;
+      sum = THCNumerics<T>::add(sum, val);
```

Comment:
wait, how are the other cases not broken then?

Reply:
it was chance that that half fixed the issue. After adding more unit tests, found that a reinterpret_cast is also required, and that int8 was also broken.